### PR TITLE
first round `geometry.voronoi_skeleton()` tests

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR:
* provide an initial round of `geometry.voronoi_skeleton(`) tests
* xref:
  * #26 
  * #25 
  * #20

### CODE REVIEW

File: sgeop/geometry.py [hunk 1]
```diff
@@ -97,33 +97,40 @@ def angle_between_two_lines(
...
-        series of geometries that shall be connected to the skeleton
+        LineStrings connected at endpoints. If ``poly`` is passed in, ``lines``
+        must be a ``geopandas.GeoSeries``.
+    poly : None | shapely.Polygon = None
+        Polygon enclosed by ``lines``.
+    snap_to : None | gpd.GeoSeries = None
+        Series of geometries that shall be connected to the skeleton.
+    max_segment_length: int = 1
+        ...
     distance : float
```
Comment (jGaboardi) [inline sgeop/geometry.py:123]: #25

File: sgeop/tests/test_geometry.py [hunk 2]
```diff
@@ -73,7 +76,7 @@ def test_not_within(self):
         assert known == observed
 
 
-class TestAngelBetween2Lines:
```
Comment (jGaboardi) [inline sgeop/tests/test_geometry.py:76]: | 😇 |

**Comment by jGaboardi** [comment]
@martinfleis Do we think I should:
* [ ] spend more time adapting tests for minimum `geopandas` 
* [ ] skip some minimum `geopandas` tests
* [ ] bump minimum `geopandas` version

**Comment by martinfleis** [comment]
I have no clue where the issue comes from but I'd bump minimum versions to whatever works. We don't need to service legacy code here.

File: README.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # `sgeop`: Street Geometry Processing Toolkit
 
-[![Continuous Integration](https://github.com/uscuni/sgeop/actions/workflows/testing.yml/badge.svg)](https://github.com/uscuni/sgeop/actions/workflows/testing.yml)
+[![Continuous Integration](https://github.com/uscuni/sgeop/actions/workflows/testing.yml/badge.svg)](https://github.com/uscuni/sgeop/actions/workflows/testing.yml) [![codecov](https://codecov.io/gh/uscuni/sgeop/branch/main/graph/badge.svg?token=VNn0WR5JWT)](https://codecov.io/gh/uscuni/sgeop)
```
Comment (jGaboardi) [inline README.md:3]: I have requested access to configure `codecov.io` for the `uscuni/sgeop` repo. Not sure if a notification was sent to you or not @martinfleis ?

Comment (martinfleis) [reply README.md:3]: approved

Comment (jGaboardi) [reply README.md:3]: Still nothing showing up. 99% sure it's because I am not listed as a member of the [`uscuni`](https://github.com/uscuni) organization (and that's why we couldn't work it out for `simplification` either). Can you add me there and we'll see if that takes care of it?

Comment (jGaboardi) [reply README.md:3]: Screenshot 2024-10-06 at 5 01 28 PM

**Comment by jGaboardi** [comment]
Something in this PR leads to the following failure:

```
 1038 # skeletonize
-> 1039 skel, _ = voronoi_skeleton(
   1040     skeletonization_input,
   1041     cluster_geom,
   1042     snap_to=False,
   1043     max_segment_length=max_segment_length,
   1044     limit_distance=1e-4,
   1045     consolidation_tolerance=consolidation_tolerance,
   1046 )
   1048 # if we used only segments, we need to remove dangles
   1049 if len(connection) == 1:
```

Need to be running a full `sgeop.simplify()` call in testing (even if not explicitly testing anything yet), just to make sure it passes.

**Comment by jGaboardi** [comment]
OK, @martinfleis – this is now ready for review.

File: README.md (hunk 3 above)
Comment (martinfleis) [reply README.md:3]: I think this is done. If not, please send me the link where I need to click something.

File: sgeop/geometry.py (hunk 1 above)
Comment (martinfleis) [reply sgeop/geometry.py:121]: Additional vertices will be added so that all line segments are no longer than this value. Must be greater than 0.

File: sgeop/geometry.py [hunk 4]
```diff
@@ -133,6 +138,8 @@ def voronoi_skeleton(
     if buffer is None:
         buffer = max_segment_length * 20
     if not poly:
+        if not isinstance(lines, gpd.GeoSeries):
+            lines = gpd.GeoSeries(lines)
```
Comment (martinfleis) [inline sgeop/geometry.py:142]: Why?

File: sgeop/tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    | geopandas.GeoSeries
+)
+
+
+def polygonize(
+    collection: line_collection, as_geom: bool = True
+) -> shapely.Polygon | geopandas.GeoSeries:
+    """Testing helper -- Create polygon from collection of lines."""
+    if isinstance(collection, pandas.Series | geopandas.GeoSeries):
+        _poly = geopandas.GeoSeries(collection).polygonize().buffer(0)
```
Comment (martinfleis) [inline sgeop/tests/conftest.py:29]: Why `buffer(0)`?

File: sgeop/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        if as_geom:
+            return _poly.squeeze()
+        else:
+            return _poly
+    else:
+        return shapely.polygonize(collection).buffer(0)
```
Comment (martinfleis) [inline sgeop/tests/conftest.py:35]: Same here

File: sgeop/tests/conftest.py (hunk 5 above)
Comment (jGaboardi) [reply sgeop/tests/conftest.py:29]: Without `buffer(0)` I get the following errors:

```
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square[array-True-0.1] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square_snap_to[list-True-0.001] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square[list-True-0.001] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square_snap_to[array-True-0.1] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
```

File: sgeop/geometry.py (hunk 4 above)
Comment (jGaboardi) [reply sgeop/geometry.py:142]: In `L143`,  the `lines` object must have a `total_bounds` attribute. The [current docstrings](https://github.com/uscuni/sgeop/blob/cf1774d85e6eb3628d9acf0ad583c8f001349fc5/sgeop/geometry.py#L114) state that something array-like is acceptable to pass in. If that's the case, `lines` must be converted to (at least) a `GeoSeries`, no?

File: sgeop/tests/conftest.py (hunk 5 above)
Comment (martinfleis) [reply sgeop/tests/conftest.py:29]: What if tou use make_valid instead?

Comment (jGaboardi) [reply sgeop/tests/conftest.py:29]: > What if tou use make_valid instead?

Good idea! Let me try that.

Comment (jGaboardi) [reply sgeop/tests/conftest.py:29]: Seems like `.buffer(0)` is actually not needed here -- neither is `.make_valid()`

File: sgeop/tests/conftest.py (hunk 6 above)
Comment (jGaboardi) [reply sgeop/tests/conftest.py:35]: `shapely.make_valid()` is not fixing the problem locally (perhaps solution in current [dev](https://shapely.readthedocs.io/en/latest/reference/shapely.make_valid.html#shapely-make-valid)) -- still need `.buffer(0)`

Comment (jGaboardi) [reply sgeop/tests/conftest.py:35]: Without `.buffer(0)` still seeing:

```
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square[array-True-0.1] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square_snap_to[list-True-0.001] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square[list-True-0.001] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
FAILED sgeop/tests/test_geometry.py::TestVoronoiSkeleton::test_square_snap_to[array-True-0.1] - AttributeError: 'GeometryCollection' object has no attribute 'exterior'
```

**Review by martinfleis — APPROVED** [review]
Alright, let's give this a go.
